Vm disk feature (#412)
terraform-providers/terraform-provider-vcd: website/docs/r/vapp_vm.html.markdown

  depends_on = ["vcd_vapp.web"]
}
```

## Example Usage (Override Template Disk)

> 

Reviewer: I suggest adding a sentence here:
"This example shows how to [change VM template's disk properties](link to Override template Disk section) when the VM is created."

Author: done

  depends_on = ["vcd_vapp.web"]
}

> resource "vcd_vapp_vm" "internalDiskOverride" {

Reviewer: The example became huge. Let's split it at this point into two examples, like we have in EG:
https://www.terraform.io/docs/providers/vcd/r/edgegateway.html
This example's header section could be "Example Usage (Override Template Disk)"

Author: added

Reviewer: Thanks! Let's just remove the network and vApp part from it to keep it focused on VM (we already have a sample with vApp above):
```
resource "vcd_network_direct" "net" {
  name             = "net"
  external_network = "corp-network"
}

resource "vcd_vapp" "web" {
  name = "web"
  depends_on = ["vcd_network_direct.net"]
}
```

Author: done

  vapp_name     = "${vcd_vapp.web.name}"
  name          = "internalDiskOverride"
  catalog_name  = "Boxes"
  template_name = "lampstack-1.10.1-ubuntu-10.04"
  memory        = 2048
  cpus          = 2
  cpu_cores     = 1

  override_template_disk {
    bus_type         = "paravirtual"
    size_in_mb       = "22384"
    bus_number       = 0
    unit_number      = 0
    iops             = 0

>     thin_provisioned = true

Reviewer 2: With VCD 10 in HTML5 GUI you cannot set thin_provisioned.
And you set the thin_provisioned setting as global ORG-VDC option. You cannot overwrite it per VM.

So thin_provisioned should be a "read" attribute only, not an argument.
Just tested it: Tested true and false, but result is the default ORG-VDC setting:

```hcl
resource "vcd_vapp" "vapp" {
  name = var.hostname
}

resource "vcd_vapp_vm" "instance" {
  name          = var.hostname
  computer_name = var.hostname
  vapp_name     = vcd_vapp.vapp.name
  catalog_name  = var.image_catalog
  template_name = var.image_string
  cpus          = var.cpu
  cpu_cores     = var.core
  memory        = var.memory

  override_template_disk {
    bus_type    = var.bus_type
    size_in_mb  = "40960"
    bus_number  = 0
    unit_number = 0
    thin_provisioned = false
  }

  network {
    type               = "org"
    name               = var.network
    ip_allocation_mode = var.private_ip == null ? "POOL" : "MANUAL"
    ip                 = var.private_ip
    is_primary         = true
  }
}
```

`terraform show`
```
resource "vcd_vapp_vm" "instance" {
    accept_all_eulas               = true
    catalog_name                   = "asys-linux"
    computer_name                  = "test1"
    cpu_cores                      = 1
    cpus                           = 1
    expose_hardware_virtualization = false
    href                           = "https://access.devcloud.arvato-systems.de/api/vApp/vm-6c6bdee2-8c3e-475a-bba2-ad0d8f99a23a"
    id                             = "urn:vcloud:vm:6c6bdee2-8c3e-475a-bba2-ad0d8f99a23a"
    internal_disk                  = [
        {
            bus_number       = 0
            bus_type         = "paravirtual"
            disk_id          = "2000"
            iops             = 0
            size_in_mb       = 40960
            storage_profile  = "Intermediate"
            thin_provisioned = true
            unit_number      = 0
        },
    ]
[...]

    network {
[...]
    }

    override_template_disk {
        bus_number       = 0
        bus_type         = "paravirtual"
        size_in_mb       = 40960
        thin_provisioned = false
        unit_number      = 0
    }
}

```

Author: Hi, yes you right. I did test and made changes in internal disk resources. Same changes for overide_template_disk also on the way. Thank you for testing it.

Author: done

Reviewer: Looks like it was completely removed, but the ask was to keep it as read-only attribute:
> So thin_provisioned should be a "read" attribute only, not an argument.

Author: override_template_disk is not read - only input. Thin provisioned value is in read only attribute vm.internal_disk.thin_provisioned

Reviewer: Please add that override_template_disk is not read to the "Override template disk" section in the docs before.

Author: added

Reviewer: Thanks!

Polishing:
`unit_number Changes are ignored on update. This part isn't reread on refresh.`
There's a dot missing at the end of the sentence after "unit_number".

Author: done

* `guest_properties` - (Optional; *v2.5+*) Key value map of guest properties
* `description`  - (Computed; *v2.6+*) The VM description. Note: description is read only. Currently, this field has
  the description of the OVA used to create the VM

> * `override_template_disk` - (Optional; *v2.6+*) Allows to update internal disk in template before first VM boot. Disk are matched by `bus_type`, `bus_number` and `unit_number`. See [Override template Disk](#override_template_disk) below for details.

Reviewer 3: ```suggestion
* `override_template_disk` - (Optional; *v2.7+*) Allows to update internal disk in template before first VM boot. Disk are matched by `bus_type`, `bus_number` and `unit_number`. See [Override template Disk](#override_template_disk) below for details.
```

Author: updated

> * `override_template_disk` - (Optional; *v2.7+*) Allows to update internal disk in template before first VM boot. Disk are matched by `bus_type`, `bus_number` and `unit_number`. See [Override template Disk](#override_template_disk) below for details.

Reviewer: `See [Override template Disk](#override_template_disk)` anchor link doesn't work (brings back to this bullet point as opposed to the section below).

Author: works for me

Author: change, now should work for you too

  * `ip_allocation_mode=NONE` - **`ip`** field can be omitted or set to an empty string "". Empty string may be useful when doing HCL variable interpolation.

<a id="override_template_disk"></a>
## Override template disk

> 

Reviewer: 1. A general comment about the block is needed here.
2. Also, this is a good place to add an explanation when *not* to use this block and use internal disk resource instead. and a link to the internal disk resource.

Author: updated

Reviewer: > Changes are ignored on update.

This is an interesting point. Won't this cause false expectations? Imagine a user who doesn't know this, he/she updates the override_template_disk config and apply succeeds, but disk is not changed. Any thoughts on how to address this?

CC: @Didainius @dataclouder

Reviewer 2: If I understood override_template_disk correctly, changes to override_template_disk config will force a re-create of a VM with the correct parameter. The situation update config and apply without changes does not happen.

For me, it was unusual that there is a "one time shot" in VM definition. Usually you write the terraform template code, and then you manage the created objects with this code.

Author: > > Changes are ignored on update.
> 
> This is an interesting point. Won't this cause false expectations? Imagine a user who doesn't know this, he/she updates the override_template_disk config and apply succeeds, but disk is not changed. Any thoughts on how to address this?
> 
> CC: @Didainius @dataclouder

I could in code make check and throw warning if `override_template_disk` is changed

Author: > If I understood override_template_disk correctly, changes to override_template_disk config will force a re-create of a VM with the correct parameter. The situation update config and apply without changes does not happen.
> 
> For me, it was unusual that there is a "one time shot" in VM definition. Usually you write the terraform template code, and then you manage the created objects with this code.

Yes we understand that. Also issue is here that in UI it's the same and users requested such functionality - increase template disk size on VM creation.

Reviewer 2: Don't know how difficult it is, but in theory, you can import the data of the template disks after VM is created. 
Let me try to explain with more details:
My idea is `override_template_disk` have 2 roles. 
If "Create", the data is used to fill the dialog known from UI. 
But in the end of `resourceVcdVAppVmCreate` in `resource_vcd_vapp_vm.go` you can import the disks (all disks present are template disks, since other internal disks will be build later after resourceVcdVAppVmCreate finished).
Now you are able to update the disk via `resourceVcdVAppVmUpdate` if `override_template_disk` has been changed.

From a user point of view this is easy to understand:
- If I want to modify the boot disk that is provided by the catalog_item, I need to modify the VM (because you will get it automatically with the VM)
- If I want to manage additional internal disks, I need to use `vcd_vm_internal_disk`

Author: Hi @buec011, thank you for your ideas. We tried many ways and various designs. And the end we have such implementation which has lowest limitations and works. Easiest way would have internal disk as typeSet in VM resource, but we couldn't due various issues and limitations:
* typeSet doesn't work well with different types and attributes of types in plan stage. Messup a lot - result user don't understand what will be final result. For e.g. change one internal disk - it will show that will change all of them, though in reality code will change only one.
* typeSet - using typeSet and disks isn't possible to implement normal update behaviour as you need keep track of diskID which is computed field and will be lost when any change is done in 
HCL as typeSet haschode will change and doesn't match in existing state file.
Use TypeList also isn't possible due that typeList is for ordered list and we don't have that.
As result separate resource allows us to implement the biggest part functionality and with little caveat.

  * `ip_allocation_mode=NONE` - **`ip`** field can be omitted or set to an empty string "". Empty string may be useful when doing HCL variable interpolation.

<a id="override_template_disk"></a>
## Override template disk
Allows to update internal disk in template before first VM boot. Disk are matched by `bus_type`, `bus_number` and `unit_number`

> Changes are ignored on update. To manage internal disk later please use [`vcd_vm_internal_disk`](/docs/providers/vcd/r/vm_internal_disk.html) resource. 

Reviewer: Need a comment that VDC must be **not** fast provisioned for this to work.

Author: added

Reviewer: Let's polish the phrasing to:
"Managing disks in VM is possible only when VDC fast provisioned is disabled."
Also, as it is pretty important limitation, what about putting it into a yellow note block?

Author: done

  * `ip_allocation_mode=NONE` - **`ip`** field can be omitted or set to an empty string "". Empty string may be useful when doing HCL variable interpolation.

<a id="override-template-disk"></a>
## Override template disk

> Allows to update internal disk in template before first VM boot. Disk are matched by `bus_type`, `bus_number` and `unit_number`

Reviewer 4: "Disk is" or "disks are" ....
```suggestion
Allows to update internal disk in template before first VM boot. Disk is matched by `bus_type`, `bus_number` and `unit_number`
```

Author: fixed

}
```

## Attribute Reference

The following additional attributes are exported:

> * `internal_disk` - (*v2.6+*) A block provides internal disk of VM details. See [Internal Disk](#internalDisk) below for details.

Reviewer 3: ```suggestion
* `internal_disk` - (*v2.7+*) A block providing internal disk of VM details. See [Internal Disk](#internalDisk) below for details.
```

Author: updated

<a id="internalDisk"></a>
## Internal disk

> * `disk_id` - (*v2.6+*) Specifies a unique identifier for this disk in the scope of the corresponding VM.

Reviewer 3: Change all `2.6` to `2.7` in this block

Author: updated
